### PR TITLE
OAK-10595 : explicit invalidation journal entry after collsion rollback

### PR DESCRIPTION
**Pull request by stefan-egli**

### CODE REVIEW

**Comment by stefan-egli** [comment]
Added a description of the suggested fix in [this comment](https://issues.apache.org/jira/browse/OAK-10595?focusedCommentId=17810500&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17810500)

**Review by Joscorbe — APPROVED** [review]
Looks good, I wonder if they could be more issues in different places where the root cause is a not properly updated/invalidated cache entry.

**Review by rishabhdaim — APPROVED** [review]

**Comment by stefan-egli** [comment]
> I wonder if they could be more issues in different places where the root cause is a not properly updated/invalidated cache entry.

It could affect whenever a written revision is later reverted - which is what the rollback is doing. I'll revisit for other potential rollback cases that wouldn't go through the now fixed code.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java [hunk 1]
```diff
@@ -2640,20 +2687,25 @@ private void invalidatePaths(@NotNull Set<Path> pathsToInvalidate) {
             // nothing to do
             return;
         }
+        invalidatePaths0(pathsToInvalidate, "split",
+                " Will be retried with next background split operation.");
+    }
+
+    private void invalidatePaths0(@NotNull Iterable<Path> pathsToInvalidate, String type,
```
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: this cries for a better method name :-)

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: `invalidatePaths1` ? ;)

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: (open for suggestions!)

**Comment by stefan-egli** [comment]
still planning on generalizing for any DocumentStore as [discussed here](https://issues.apache.org/jira/browse/OAK-10595?focusedCommentId=17810916&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17810916)

**Comment by stefan-egli** [comment]
@reschke RDB test added [here](https://github.com/apache/jackrabbit-oak/pull/1284/commits/d9566b56536dc6f9511458320ddf1b18ba1ba24a)

**Comment by stefan-egli** [comment]
toggle introduced in [this commit](https://github.com/apache/jackrabbit-oak/pull/1284/commits/ad9f90ca178157ecfb9e0b82db221341c23d97eb)

**Review by rishabhdaim — APPROVED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java (hunk 1 above)
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: What about `createInvalidationJournalEntry`?

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: thx, renamed in https://github.com/apache/jackrabbit-oak/pull/1284/commits/e0bd8a1c1ec567ab551c787fea9f67a052fe6b44

**Pull request merged**